src/System.Net.Security/tests/StressTests/SslStress/Utils/PipeExtensions.cs

```diff
@@ -0,0 +1,93 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Buffers;
+using System.IO;
+using System.IO.Pipelines;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace SslStress.Utils
+{
+    public static class PipeExtensions
+    {
+        // Adapted from https://devblogs.microsoft.com/dotnet/system-io-pipelines-high-performance-io-in-net/
+        public static async Task ReadLinesUsingPipesAsync(this Stream stream, Func<ReadOnlySequence<byte>, Task> callback, CancellationToken token = default, char separator = '\n')
+        {
+            var pipe = new Pipe();
+            PipeWriter writer = pipe.Writer;
+            PipeReader reader = pipe.Reader;
+
+            try
+            {
+                await TaskExtensions.WhenAllCancelOnFirstException(token, FillPipeAsync, ReadPipeAsync);
+            }
+            catch when (token.IsCancellationRequested)
```

Comment:
Can this be narrowed down to avoid ignoring unintended cancellations? Not sure off hand if `WhenAllCancelOnFirstException` does this correctly for this to even work.

`catch(OperationCancelledException ex) when(ex.CancellationToken == token)`

Reply:
It won't work in this case, because `WhenAllCancelOnFirstException` uses a linked cancellation token (which can get canceled if one of the sibling tasks throws).